pilot/pkg/networking/grpcgen/cds.go

```diff
@@ -0,0 +1,270 @@
+	var defaultCluster *clusterv3.Cluster
+	if b.filter.Contains(b.defaultClusterName) {
+		defaultCluster = edsCluster(b.defaultClusterName)
+	}
+
+	subsetClusters := b.applyDestinationRule(defaultCluster)
+	out := make([]*clusterv3.Cluster, 0, 1+len(subsetClusters))
+	if defaultCluster != nil {
+		out = append(out, defaultCluster)
+	}
+	return append(out, subsetClusters...)
+}
+
+// edsCluster creates a simple cluster to read endpoints from ads/eds.
+func edsCluster(name string) *clusterv3.Cluster {
+	return &clusterv3.Cluster{
+		Name:                 name,
+		ClusterDiscoveryType: &clusterv3.Cluster_Type{Type: clusterv3.Cluster_EDS},
+		EdsClusterConfig: &clusterv3.Cluster_EdsClusterConfig{
+			ServiceName: name,
+			EdsConfig: &envoy_config_core_v3.ConfigSource{
+				ConfigSourceSpecifier: &envoy_config_core_v3.ConfigSource_Ads{
+					Ads: &envoy_config_core_v3.AggregatedConfigSource{},
+				},
+			},
+		},
+	}
+}
+
+// applyDestinationRule mutates the default cluster to reflect traffic policies, and returns a set of additional
+// subset clusters if specified by a destination rule
+func (b *clusterBuilder) applyDestinationRule(defaultCluster *clusterv3.Cluster) (subsetClusters []*clusterv3.Cluster) {
+	if b.svc == nil || b.port == nil {
+		return nil
+	}
+
+	// resolve policy from context
+	destinationRule := v1alpha3.CastDestinationRule(b.push.DestinationRule(b.node, b.svc))
+	trafficPolicy := v1alpha3.MergeTrafficPolicy(nil, destinationRule.TrafficPolicy, b.port)
+
+	// setup default cluster
+	b.applyPolicy(defaultCluster, trafficPolicy)
+
+	// subset clusters
+	if len(destinationRule.Subsets) > 0 {
+		subsetClusters = make([]*clusterv3.Cluster, 0, len(destinationRule.Subsets))
+		for _, subset := range destinationRule.Subsets {
+			subsetKey := subsetClusterKey(subset.Name, string(b.hostname), b.portNum)
+			if !b.filter.Contains(subsetKey) {
+				continue
+			}
+			c := edsCluster(subsetKey)
+			trafficPolicy := v1alpha3.MergeTrafficPolicy(trafficPolicy, subset.TrafficPolicy, b.port)
+			b.applyPolicy(c, trafficPolicy)
+			subsetClusters = append(subsetClusters, c)
+		}
+	}
+
+	return
+}
+
+// applyPolicy mutates the give cluster (if not-nil) so that the given merged traffic policy applies.
+func (b *clusterBuilder) applyPolicy(c *clusterv3.Cluster, trafficPolicy *networking.TrafficPolicy) {
+	// cluster can be nil if it wasn't requested
+	if c == nil || trafficPolicy == nil {
+		return
+	}
+	b.applyTLS(c, trafficPolicy)
+	b.applyLoadBalancing(c, trafficPolicy)
+	// TODO status or log when unsupported features are included
+}
+
+func (b *clusterBuilder) applyLoadBalancing(c *clusterv3.Cluster, policy *networking.TrafficPolicy) {
+	switch policy.LoadBalancer.GetSimple() {
+	case networking.LoadBalancerSettings_ROUND_ROBIN:
+	// ok
+	default:
+		log.Warnf("cannot apply LbPolicy %s to %s", policy.LoadBalancer.GetSimple(), b.node.ID)
+	}
+
+	// TODO https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md
```

Comment:
so once ring hash load balancing gets implemented in go, we will allow it to be applied?

Reply:
gRPC does support it now, I just have a TODO to actually apply it in our CDS